# Disciplina: Introdução a programação para geocientistas

# Teste 02

Este teste foi pensado para que se utilize, em todas as questões, as estruturas de repetição (`for` ou `while`). Em algumas será necessário utilizar as estruturas de decisão (`if/elif/else`).

### Questão 1 (1 ponto)



Escreva um programa que receba uma lista qualquer e calcule o número de elementos da lista. **NÃO** utilizar a função `len()` ou `count()`.

In [6]:
lista = [1, 5, 10, 15, 20, 25, 30]
cont = 0

for i in lista:
    cont += 1

print ('Há', cont, 'elementos na lista')

Há 7 elementos na lista


### Questão 2 (1 ponto)

Escreva um programa para converter temperaturas de Fahrenheit para Celsius usando a fórmula abaixo:

$$ C = (5 \times (F - 32))/9 $$

Em que $C$ é a temperatura em graus Celcius e $F$ é a temperatura em graus Fahrenheit. Para fazer a questão, primeiro crie uma lista de temperaturas em graus Fahrenheit usando a função `random.sample()`. A lista deve conter 100 números aleatórios que variem entre 10 a 150.

Uma vez gerada a lista de temperaturas convertidas em graus Celcius, o programa também deve informar quantos valores correspondem a temperaturas abaixo de zero. 

In [40]:
import random
lista = random.sample (range (10,150),100)
print ('Lista de graus em Fahrenheit: ', lista)
nova_lista = []
cont = 0

for i in lista:
    C = (5*(i - 32))/9
    nova_lista.append (C)    
print ('Lista de graus convertidos para Celsius: ', nova_lista)

for i in nova_lista:
    if i < 0:
        cont += 1
print (cont, 'valores correspondem a temperaturas abaixo de zero')

Lista de graus em Fahrenheit:  [60, 75, 96, 109, 120, 12, 94, 123, 117, 132, 139, 114, 98, 131, 37, 22, 50, 88, 15, 43, 59, 101, 55, 70, 65, 107, 144, 108, 14, 64, 58, 25, 54, 33, 67, 16, 46, 17, 39, 42, 18, 90, 102, 76, 20, 68, 57, 83, 141, 113, 81, 128, 48, 77, 31, 56, 87, 99, 110, 38, 133, 105, 149, 84, 119, 129, 118, 124, 62, 122, 135, 89, 142, 28, 134, 127, 103, 27, 78, 126, 23, 52, 79, 74, 95, 130, 73, 147, 143, 92, 26, 35, 63, 10, 80, 44, 13, 72, 11, 40]
Lista de graus convertidos para Celsius:  [15.555555555555555, 23.88888888888889, 35.55555555555556, 42.77777777777778, 48.888888888888886, -11.11111111111111, 34.44444444444444, 50.55555555555556, 47.22222222222222, 55.55555555555556, 59.44444444444444, 45.55555555555556, 36.666666666666664, 55.0, 2.7777777777777777, -5.555555555555555, 10.0, 31.11111111111111, -9.444444444444445, 6.111111111111111, 15.0, 38.333333333333336, 12.777777777777779, 21.11111111111111, 18.333333333333332, 41.666666666666664, 62.22222222222222, 42.222

### Questão 3 (1,5 pontos)



Seja $\mathbf t$ um vetor que contém valores de tempo, calcule $\mathbf y$ de acordo com a equação a seguir:
$$
\mathbf y = m_{1} + m_{2}\mathbf t - \frac{1}{2}m_{3}\mathbf t² \: ,
$$

em que 

$$\mathbf{t} = \left[
\begin{array}{c}
1, 2, 3, 4, 5, 6, 7, 8, 9, 10
\end{array}
\right]$$

e

$$m_1 = 10; m_2 = 100; m_3 = 9.8$$

Resposta esperada: $$\mathbf{y} = \left[ 
\begin{array}{c}
105.1 ,190.4, 265.9, 331.6, 387.5, 433.6, 469.9, 496.4, 513.1, 520.
\end{array}
\right]$$

In [52]:
t = [1,2,3,4,5,6,7,8,9,10]
m1 = 10
m2 = 100
m3 = 9.8
list = []

for i in t:
    y = m1 + m2*i-1/2*m3*i**2
    list.append (y)

print ('y =', list)

y = [105.1, 190.4, 265.9, 331.6, 387.5, 433.6, 469.9, 496.4, 513.0999999999999, 520.0]


### Questão 4 (1,5 pontos)

Escreva um programa que gere $n$ estações igualmente espaçadas no intervalo de `a` a `b`, sendo `a` o início e `b` o fim. Guarde as coordenadas em uma lista `coordenadas`.

Por exemplo: Caso deseje que a lista tenha 10 estações entre 10 e 100, teremos 9 intervalos com espaçamento igual a 10. Nesse caso, o resultado final seria uma lista de coordenadas = [10, 20, 30, 40, 50, 60, 70, 80, 90, 100]

*Dica*: Uma lista com $n$ estações, teremos $n-1$ intervalos, o espaçamento é dado por $ h = (b-a)/(n-1)$. As coordenadas podem ser geradas pela fórmula: $ x_i = a + i h$, sendo $i = 0, 1, ..., n+1.$

In [27]:
a = int(input('digite o início do intervalo: '))
b = int(input('digite o fim do intervalo: '))
n = int (input ('insira a quantidade de estações: ')) 

h = (b-a)/(n - 1)   # espaçamento -- distância
estacoes = []

for i in range(n):
    xi = a + i*h
    estacoes.append(xi)

print(estacoes)

digite o início do intervalo: 10
digite o fim do intervalo: 100
insira a quantidade de estações: 10
[10.0, 20.0, 30.0, 40.0, 50.0, 60.0, 70.0, 80.0, 90.0, 100.0]


## Anomalia gravimétrica

Uma anomalia gravimétrica é a diferença entre a aceleração da gravidade observada em um determinado local (geralmente um ponto na superfície da Terra), e o valor correspondente previsto a partir de um modelo do campo gravitacional do planeta. Normalmente, o modelo baseia-se em premissas simplificadas, tais como a de que o planeta é aproximado por elipsóide de revolução que contempla as dimensões, massa e rotação da Terra.

A diferença entre o valor da aceleração da gravidade observada ($g_{obs}$) e a aceleração da gravidade teórica ($\gamma$) é obtida através da fórmula: 

$$\Delta g=g_{obs}-\gamma$$ 

em que $\Delta g$ é a anomalia gravimétrica. As anomalias gravimétricas descrevem as variações locais do campo de gravidade em relação ao campo gravitacional previsto pelo modelo ($\gamma)$. Um local com uma anomalia positiva exibe mais gravidade do que o previsto pelo modelo sugerindo a presença de um excesso de massa, enquanto uma anomalia negativa exibe um valor mais baixo do que o previsto - sugerindo um déficit de massa. Estas anomalias são, portanto, de substancial interesse geofísico e geológico.

**Anomalia Ar livre**

De forma bastante simplificada, é a anomalia de gravidade medida após a aplicação de uma correção Ar-livre para compensar a elevação em que é feita uma medição. Para isso, ajustam-se estas medidas de gravidade ao que teria sido medido a um nível de referência, que é normalmente tomado como o nível médio do mar ou o geóide. O cálculo da anomalia é dado pela fórmula a seguir: 

$$ \Delta g_{ar-livre}=g_{obs}-\gamma-(-0,3086h) $$

Onde 0,3086 é o gradiente de gravidade normal expresso em mGal/m e h é a elevação do ponto onde foi realizada a medida da aceleração da gravidade, em metros.

**Anomalia Bouguer** 

É a anomalia corrigida para compensar, não só a elevação como a correção Ar-livre, mas também as massas presentes entre o ponto de observação e a superfície de referência. Ao aplicar esta correção, a anomalia restante é representativa sobretudo de variações de densidade em subsuperfície. A anomalia de Bouguer é dada por:

$$ \Delta g_{bouguer}=g_{obs}-\gamma- (−0.3086 h +2\pi G \rho \times10^{5} h ) $$

em que $G = 6.674184 \times 10^{-11}$ é a constante gravitacional e $\rho = 2670$ é a densidade de referência do continente em g/cm³.

### Questão 5 (1,5 pontos)

A partir dos vetores de gravidade observada ($g_{obs}$), gravidade teórica ($\gamma$) e altitudes ($h$) fornecidos abaixo, calcule a anomalia Ar-livre utilizando a fórmula acima correspondente.

*Resposta esperada* 

$\Delta g_{ar-livre} = [29.17, 32.17, 34.34, 38.66, 41.33, 29.97, 29.42, 18.02, 13.16, 21.32, 23.38, 39.02, 52.39, 9.52, 17.19, 27.07, 67.91, 1.2, 39.65, 8.33, 53.8, 85.39, 20.36, 2.02, 3.66, -2.32, 11.55, 16.94, 19.94, 47.25]$

In [29]:
# Gravidade observada

g_obs = [978327.75895173, 978356.13786764, 978366.69919956, 978373.62929348, 978382.09717987, 
         978380.20431098, 978380.77447999, 978369.46532367, 978363.12556894, 978372.82134202, 
         978374.88115842, 978390.52501546, 978403.82877837, 978360.9993912 , 978368.69042058, 
         978378.45457184, 978419.41702798, 978352.70233265, 978391.15784458, 978359.8325173, 
         978405.29508957, 978436.89961504, 978371.8668729 , 978353.49779395, 978355.16050153, 
         978349.18970325, 978363.05588538, 978368.4333742, 978371.43210584, 978398.75068886]

# Gravidade teórica

gamma = [978351.50543132, 978351.50543132, 978351.50543132, 978351.50543132,
       978351.50543132, 978351.50543132, 978351.50543132, 978351.50543132,
       978351.50543132, 978351.50543132, 978351.50543132, 978351.50543132,
       978351.50543132, 978351.50543132, 978351.50543132, 978351.50543132,
       978351.50543132, 978351.50543132, 978351.50543132, 978351.50543132,
       978351.50543132, 978351.50543132, 978351.50543132, 978351.50543132,
       978351.50543132, 978351.50543132, 978351.50543132, 978351.50543132,
       978351.50543132, 978351.50543132]

# Altitude

h = [171.48,  89.24,  62.04,  53.6 ,  34.8 ,   4.12,   0.  ,   0.  ,
         5.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,
         0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ,
         0.  ,   0.  ,   0.  ,   0.  ,   0.  ,   0.  ]

In [30]:
# Δ𝑔𝑎𝑟−𝑙𝑖𝑣𝑟𝑒 = 𝑔𝑜𝑏𝑠−𝛾−(−0,3086ℎ)   # gobs-y = gravidade observada  #y = aceleração da gravidade teórica
# FÓRMULA: gar_livre = g_obs - gamma - (-0.3086)

gar_livre = []

for i in range(len(g_obs)):
    gar_livre_i = g_obs[i] - gamma[i] - (-0.3086 * h[i])
    gar_livre.append(gar_livre_i)
    
print ('Δ𝑔𝑎𝑟−𝑙𝑖𝑣𝑟𝑒 =', gar_livre)

Δ𝑔𝑎𝑟−𝑙𝑖𝑣𝑟𝑒 = [29.172248409925835, 32.171900319952826, 34.33931223992934, 38.66482215990409, 41.33102854995217, 29.97031165999078, 29.269048669957556, 17.959892349899746, 13.163137619902379, 21.31591069989372, 23.375727099948563, 39.01958413992543, 52.32334704999812, 9.493959879968315, 17.184989259927534, 26.94914051995147, 67.91159665992018, 1.1969013299094513, 39.65241325995885, 8.327085979981348, 53.78965824993793, 85.39418371999636, 20.361441579996608, 1.992362629971467, 3.655070209992118, -2.3157280700979754, 11.550454059964977, 16.927942879963666, 19.92667451989837, 47.24525753990747]


### Questão 6 (1,5 pontos)

A partir dos vetores de gravidade observada ($g_{obs}$), gravidade teórica ($\gamma$) e altitudes ($h$) fornecidos na questão anterior, calcule a anomalia Bouguer utilizando a fórmula dada anteriormente.

*Resposta esperada*

$\Delta g_{bouguer} = [9.97, 22.18, 27.39, 32.66, 37.43, 29.51, 29.37, 18.  , 12.6 ,
       21.32, 23.38, 39.02, 52.36,  9.51, 17.19, 27.03, 67.91,  1.2 ,
       39.65,  8.33, 53.79, 85.39, 20.36,  2.01,  3.66, -2.32, 11.55,
       16.93, 19.93, 47.25]$

In [41]:
# Δ𝑔𝑏𝑜𝑢𝑔𝑢𝑒𝑟 = 𝑔ar_livre + (2 𝜋 𝐺 𝜌 × 10^5 ℎ) | constante gravitacional - 𝜌 = 2670 | 𝐺=6.674184 × 10**(-11)
import numpy as np

𝐺 = 6.674184 * 10 **-11
p = 2670
pi = np.pi
g_bouguer = []

for i in range(len(g_obs)):
    g_bouguer_i = gar_livre[i] + ((2*pi*G*p) * (10**5 * h[i]))
    g_bouguer.append(g_bouguer_i)
    
print ('Δ𝑔𝑏𝑜𝑢𝑔𝑢𝑒𝑟 = ', g_bouguer)

Δ𝑔𝑏𝑜𝑢𝑔𝑢𝑒𝑟 =  [48.37231699548025, 42.1638184478679, 41.28573313477303, 44.666243178015336, 45.22747353932291, 30.431614917353063, 29.269048669957556, 17.959892349899746, 13.722971670099323, 21.31591069989372, 23.375727099948563, 39.01958413992543, 52.32334704999812, 9.493959879968315, 17.184989259927534, 26.94914051995147, 67.91159665992018, 1.1969013299094513, 39.65241325995885, 8.327085979981348, 53.78965824993793, 85.39418371999636, 20.361441579996608, 1.992362629971467, 3.655070209992118, -2.3157280700979754, 11.550454059964977, 16.927942879963666, 19.92667451989837, 47.24525753990747]


### Questão 7 (2 pontos)

Faça um programa que, dada as listas acima (g_obs, gamma e h), o usuário decida se deseja calcular a anomalia Ar-livre ou a anomalia Bouguer e retorne (imprima) o resultado correto. Crie uma opção para caso o usuário digite uma entrada inválida. 

*Dica: Utilize as formulas dadas no enunciado e aproveite os comandos dos programas que fez nas questões 5 e 6.*

In [33]:
calculo = input('Qual anomalia deseja calcular? Digite "ar" para Anomalia Ar-livre ou "bouguer" para Anomalia Bouguer: ')

if calculo == 'ar':
    print ('A anomalia ar é igual a:', gar_livre)
elif calculo == 'bouguer':
    print ('A anomalia Bouguer é igual a: ', g_bouguer)
else :
    print ('entrada inválida')

Qual anomalia deseja calcular? Digite "ar" para Anomalia Ar-livre ou "bouguer" para Anomalia Bouguer: nada
entrada inválida
